[Profiler] Unify the device(CUDA, XPU, PrivateUse1) in torch profiler post processing

zejun-chen opened the pull request:
This PR unifies the CUDA, XPU and PrivateUse1 in the torch profiler. Now CUDA, XPU and PrivateUse1 can together use string object `use_device` to distinguish each other and share one device path for calculating kineto time durations and memory statistics for post processing.

#suppress-api-compatibility-check

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @ezyang @gchanan @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @rohan-varma @aakhundov

zejun-chen committed:
[Profiler] Unify the device(CUDA, XPU, PrivateUse1) in profiler post processing
Use use_device to identify the required device

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
Hi @aaronenyeshi @valentinandrei

Could you help review this PR? It unifies the post processing functions for different devices in torch profiler and we enable XPU in it. Any comments are welcomed.

Thank you :)

zejun-chen committed:
Merge branch 'main' into zejun/add_xpu_post_processing_for_torch_profile

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
[fix] fix legacy profiler usage and FunctionEvent usage

aaronenyeshi commented on torch/autograd/profiler.py:
```diff
@@ -112,8 +112,8 @@ class profile:
     Args:
         enabled (bool, optional): Setting this to False makes this context manager a no-op.
 
-        use_cuda (bool, optional): Enables timing of CUDA events as well using the cudaEvent API.
```
Could we keep use_cuda around for a little, and mark it as deprecating. It usually takes 1-2 releases to mark an API as deprecated (we could leave it in the args, but as a no-op).

zejun-chen replied:
Sure. Thank you for suggestion.

aaronenyeshi commented on torch/autograd/profiler.py:
```diff
@@ -191,7 +191,6 @@ def __init__(
         self,
         enabled=True,
         *,
-        use_cuda=False,
```
please keep use_cuda, and add a comment that its deprecated.

aaronenyeshi commented on torch/autograd/profiler.py:
```diff
@@ -206,9 +205,8 @@ def __init__(
-        self.use_cuda = use_cuda
         self.use_device: Optional[str] = (
-            use_device if use_device != "privateuseone" else None
+            None if use_device == "privateuseone" else use_device
```
@fwenguang could privateuseone keep self.use_device="privateuseone"? Why do we set it to None?

fwenguang replied:
keep self.use_device="privateuseone" is okey to me, if so, some code below also need to be updated. For example, in some if conditional statements.

In fact, self.use_device="privateuseone" is equivalent to self.use_device=None  for our custom backend, since we will rename backend name to another string.

zejun-chen replied:
`        if use_cuda:
            warn("The attribute `use_cuda` will be deprecated soon, please use ``use_device = 'cuda'`` instead.")
            self.use_device = "cuda"
        else:
            self.use_device: Optional[str] = (
                "privateuseone" if use_device is None else use_device
            )`

zejun-chen replied:
OK, i change the code to below. When users set the use_cude, the use_device='cuda' and a warning is thrown. When users set the use_device, the use_device is set to what user set.
```
if use_cuda:
    warn("The attribute `use_cuda` will be deprecated soon, please use ``use_device = 'cuda'`` instead.")
    self.use_device = "cuda"
else:
    self.use_device: Optional[str] = (
        "privateuseone" if use_device is None else use_device
    )
```

aaronenyeshi replied:
warning for use_cuda looks good to me. Ithink for self.use_device, it should:
```
self.use_device: Optional[str] = use_device
```

aaronenyeshi commented on torch/autograd/profiler.py:
```diff
@@ -516,12 +509,12 @@ def _privateuse1_memory_usage(mem_record):
             max_evt_id = max(max_evt_id, fe.id)
             if fe.device_type == DeviceType.CPU and not fe.is_async:
-                if self.use_device:
+                if self.use_device and self.use_device not in ["cuda", "xpu"]:
```
privateuseone seems to set self.use_device=None, so this block may not be called. Maybe we should make privateuseone a valid str in use_device?

briancoutinho replied:
+1

aaronenyeshi commented on torch/autograd/profiler.py:
```diff
@@ -112,8 +112,8 @@ class profile:
 
-        use_cuda (bool, optional): Enables timing of CUDA events as well using the cudaEvent API.
-            Adds approximately 4us of overhead to each tensor operation.
+        use_device (str, optional): Enables timing of device events.
```
Could we also comment the list of valid options so far? Such as cuda, xpu, privateuseone.

briancoutinho replied:
yes we can also check for it being "in VAID_DEVICE_OPTIONS"

zejun-chen replied:
Thank you for comments.

aaronenyeshi reviewed (changes requested):
Overall, this looks great! Thank you for the change to consolidate the devices, it improves the code readability and maintainability in the future. 

A few comments:
1) Could we keep use_cuda as an argument for awhile. We need to mark it for deprecation in case anyone is currently using it.
2) I wonder why self.use_device is None for privateuseone. Why not support all valids types: cuda, xpu, privateuseone? cc @fwenguang , @NmomoN

briancoutinho commented on torch/autograd/profiler.py:
```diff
@@ -233,17 +231,18 @@ def __init__(
+
+        if self.use_device == "xpu" and not torch.xpu.is_available(): # type: ignore[attr-defined]
+            warn("XPU is not available, disabling XPU profiling")
+            self.use_device = None
```
please add a check if use_device is not a valid option here. 
``` if self.use_device not in VALID_USE_DEViCE_OPTIONS:
 ```

zejun-chen replied:
Sure. Thank you.

briancoutinho commented on torch/autograd/profiler.py:
```diff
@@ -252,14 +251,17 @@ def __init__(
 
         self.profiler_kind = ProfilerState.KINETO
-        if self.use_cuda:
+        if self.use_device == "cuda":
```
as @aaronenyeshi  suggested please keep use_cuda
```
if self.use_cuda or self.use_device == "cuda"
```

zejun-chen replied:
Sure. Thank you.

briancoutinho commented on torch/autograd/profiler.py:
```diff
@@ -315,8 +317,10 @@ def _start_trace(self):
         if not self.enabled:
             return
-        if self.use_cuda:
+        if self.use_device == "cuda":
```
as above, please also support use_cuda :)

briancoutinho commented on torch/autograd/profiler.py:
```diff
@@ -445,17 +448,10 @@ def _cpu_memory_usage(mem_record):
-            )
-
-        def _privateuse1_memory_usage(mem_record):
+        def _device_memory_usage(mem_record):
```
thanks for de-duping this!

briancoutinho commented on torch/utils/bottleneck/__main__.py:
```diff
@@ -92,7 +92,7 @@ def print_cprofile_summary(prof, sortby='tottime', topk=15):
 def run_autograd_prof(code, globs):
     def run_prof(use_cuda=False):
-        with profiler.profile(use_cuda=use_cuda) as prof:
+        with profiler.profile(use_device="cuda" if use_cuda else None) as prof:
```
probably keep this:)

zejun-chen replied:
Sure. Thank you.

briancoutinho commented on torch/autograd/profiler_util.py:
```diff
@@ -280,7 +281,11 @@ def export_stacks(self, path: str, metric: str):
             for evt in self:
                 if evt.stack and len(evt.stack) > 0:
-                    metric_value = getattr(evt, metric)
+                    metric_value = getattr(evt,
```
Could you explain why this replace is needed? thanks

zejun-chen replied:
Here replacing is because the unified `self_device_time_total` is used to calculate the metrics in the `FunctionEvent`

briancoutinho commented on torch/autograd/profiler_util.py:
```diff
@@ -395,26 +399,19 @@ class FormattedTimesMixin:
     self_cpu_time_total_str = _attr_formatter("self_cpu_time_total")
-    self_cuda_time_total_str = _attr_formatter("self_cuda_time_total")
-    self_privateuse1_time_total_str = _attr_formatter("self_privateuse1_time_total")
+    self_device_time_total_str = _attr_formatter("self_device_time_total")
```
thanks for simplifying this :)

briancoutinho reviewed (commented):
Thank you! This significantly improves the readability of this code and I appreciate the effort going into refactoring this. Overall looks good to me. Few minor comments and we can approve

zejun-chen committed:
Keep use_cuda and add VAID_DEVICE_OPTIONS

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Fix UFMT in torch/autograd/profiler_util.py

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
[Fix] CI case failure

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

gujinghui commented on torch/autograd/profiler.py:
```diff
@@ -315,8 +330,10 @@ def _start_trace(self):
+        if self.use_cuda or self.use_device == "cuda":
             torch.cuda.synchronize()
+        elif self.use_device == "xpu":
+            torch.xpu.synchronize()  # type: ignore[attr-defined]
```
"type: ignore[attr-defined]" is not needed, as torch.xpu is landed already.

zejun-chen replied:
OK. Sure.

zejun-chen committed:
[Fix] Keep use_cuda in prepare_trace

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
[Fix] remove type: ignore[attr-defined] for torch/autograd/profiler.py

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

aaronenyeshi reviewed (changes requested):
This looks great! Just a few small changes, and should be good to ship. Check the lint failures too please.

zejun-chen commented on torch/autograd/profiler.py:
```diff
@@ -207,8 +211,12 @@ def __init__(
+            )
         self.use_device: Optional[str] = (
-            use_device if use_device != "privateuseone" else None
+            "privateuseone" if use_device is None else use_device
```
OK, sure

zejun-chen committed:
Refine the code

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
[Fix] Fix CI failure

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
> This looks great! Just a few small changes, and should be good to ship. Check the lint failures too please.

Thank you for comments 👍 
The CI is passed. Could it be reviewed again?
Thank you.

aaronenyeshi reviewed (dismissed):
LGTM

aaronenyeshi commented:
@pytorchbot merge

zejun-chen committed:
[Fix] No annotate use_device twice

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
@pytorchbot merge

aaronenyeshi commented:
@pytorchbot merge

zejun-chen commented:
@pytorchbot rebase

zejun-chen commented:
Hi, @aaronenyeshi 

Thank you for approval.
For the lint error, here it is:
```
2024-04-11T18:09:18.6165543Z + ../_test-infra/tools/stronghold/bin/check-api-compatibility --base-commit=eff1e4899c7c89f8a8fc8f6ff6bed06dd8d2ec8a --head-commit=22c2d3bb75f2b8ca7419904ec5ecf344ca7ae8c2
2024-04-11T18:09:19.8837412Z From https://github.com/pytorch/pytorch
2024-04-11T18:09:19.8839169Z  * branch                    eff1e4899c7c89f8a8fc8f6ff6bed06dd8d2ec8a -> FETCH_HEAD
2024-04-11T18:09:20.0642879Z ##[warning]Function FormattedTimesMixin.cuda_time: function deleted
2024-04-11T18:09:20.0645300Z ##[warning]Function FormattedTimesMixin.privateuse1_time: function deleted
2024-04-11T18:09:20.0646984Z ##[warning]Function FunctionEvent.self_cuda_memory_usage: function deleted
2024-04-11T18:09:20.0648547Z ##[warning]Function FunctionEvent.self_privateuse1_memory_usage: function deleted
2024-04-11T18:09:20.0649998Z ##[warning]Function FunctionEvent.cuda_time_total: function deleted
2024-04-11T18:09:20.0651407Z ##[warning]Function FunctionEvent.self_cuda_time_total: function deleted
2024-04-11T18:09:20.0652911Z ##[warning]Function FunctionEvent.self_privateuse1_time_total: function deleted
2024-04-11T18:09:20.0654335Z ##[warning]Function FunctionEvent.privateuse1_time_total: function deleted
<probably uninteresting folded group, click to show>
2024-04-11T18:09:20.0699880Z ##[error]Process completed with exit code 1.
```

I reproduce the error in my local machine, it is caused by we are removing some member of the **FunctionEvent** and **FormattedTimesMixin** in the PR.

Here is the code throwing these errors:
https://github.com/pytorch/test-infra/blob/main/tools/stronghold/src/api/compatibility.py#L81-L82
It only ignores the member with '_' starts with.
https://github.com/pytorch/test-infra/blob/main/tools/stronghold/src/api/compatibility.py#L77-L78

Thank you.

zejun-chen committed:
Merge branch 'main' into zejun/add_xpu_post_processing_for_torch_profile

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
@pytorchbot merge

zejun-chen commented:
#suppress-api-compatibility-check #suppress-bc-linter

zejun-chen commented:
@pytorchbot #suppress-api-compatibility-check #suppress-bc-linter

zejun-chen commented:
@pytorchbot merge

zejun-chen commented:
@pytorchbot merge

zejun-chen commented:
@pytorchmergebot 

> Hi, @aaronenyeshi
> 
> Thank you for approval. For the lint error, here it is:
> 
> ```
> 2024-04-11T18:09:18.6165543Z + ../_test-infra/tools/stronghold/bin/check-api-compatibility --base-commit=eff1e4899c7c89f8a8fc8f6ff6bed06dd8d2ec8a --head-commit=22c2d3bb75f2b8ca7419904ec5ecf344ca7ae8c2
> 2024-04-11T18:09:19.8837412Z From https://github.com/pytorch/pytorch
> 2024-04-11T18:09:19.8839169Z  * branch                    eff1e4899c7c89f8a8fc8f6ff6bed06dd8d2ec8a -> FETCH_HEAD
> 2024-04-11T18:09:20.0642879Z ##[warning]Function FormattedTimesMixin.cuda_time: function deleted
> 2024-04-11T18:09:20.0645300Z ##[warning]Function FormattedTimesMixin.privateuse1_time: function deleted
> 2024-04-11T18:09:20.0646984Z ##[warning]Function FunctionEvent.self_cuda_memory_usage: function deleted
> 2024-04-11T18:09:20.0648547Z ##[warning]Function FunctionEvent.self_privateuse1_memory_usage: function deleted
> 2024-04-11T18:09:20.0649998Z ##[warning]Function FunctionEvent.cuda_time_total: function deleted
> 2024-04-11T18:09:20.0651407Z ##[warning]Function FunctionEvent.self_cuda_time_total: function deleted
> 2024-04-11T18:09:20.0652911Z ##[warning]Function FunctionEvent.self_privateuse1_time_total: function deleted
> 2024-04-11T18:09:20.0654335Z ##[warning]Function FunctionEvent.privateuse1_time_total: function deleted
> <probably uninteresting folded group, click to show>
> 2024-04-11T18:09:20.0699880Z ##[error]Process completed with exit code 1.
> ```
> 
> I reproduce the error in my local machine, it is caused by we are removing some members of the **FunctionEvent** and **FormattedTimesMixin** in the PR.
> 
> Here is the code throwing these errors by the check function: https://github.com/pytorch/test-infra/blob/main/tools/stronghold/src/api/compatibility.py#L81-L82
> 
> This check only ignores the member with '_' starts with, for example, `classA._member` https://github.com/pytorch/test-infra/blob/main/tools/stronghold/src/api/compatibility.py#L77-L78
> 
> I will kick off a issue to the https://github.com/pytorch/test-infra/tree/main and make lint happy soon. Here is the issue: [pytorch/test-infra#5093](https://github.com/pytorch/test-infra/issues/5093)
> 
> Thank you.

Hi, @aaronenyeshi 

The above lint error is fixed by adding the label: suppress-api-compatibility-check.
Now after i rebase the main branch, here is a new error shown as below:
![image](https://github.com/pytorch/pytorch/assets/74231238/b7c47e3e-8e69-41ea-a237-7353ac49ab55)
Could you import to sync the PR?

Thank you.

gujinghui commented:
@pytorchbot rebase

fwenguang commented on torch/profiler/profiler.py:
```diff
@@ -125,9 +127,13 @@ def __init__(
+            self.use_device = "cuda"
+        elif ProfilerActivity.XPU in self.activities:
+            self.use_device = "xpu"
+        else:
```
Hi @zejun-chen , after merging the main branch, we need `elif ProfilerActivity.PrivateUse1 in self.activities:` here, and delele the similar if statement in prepare_trace function. https://github.com/pytorch/pytorch/pull/123247/files#diff-f021e4bc3d3597b11937c8e8688d2a0053c0b542c8a1628326d77ef852e3e682R153-R154. Otherwise, use_device will not be set to xpu or cuda.

zejun-chen replied:
Hi, @fwenguang 

Thank you for comments. Does the below code change satisfy the requirement? You will register the `ProfilerActivity.PrivateUse1` to the activities, right?
```
        if ProfilerActivity.CUDA in self.activities:
            self.use_device = "cuda"
        elif ProfilerActivity.XPU in self.activities:
            self.use_device = "xpu"
        elif ProfilerActivity.PrivateUse1 in self.activities:
            self.use_device = _get_privateuse1_backend_name()
        else:
            # raise error
```
Thank you.

fwenguang replied:
Yes, this change is okay. Thank you.

zejun-chen committed:
[Fix] pass use_device to the prof.profile in perpare_trace

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

gujinghui commented:
@aaronenyeshi

We are unable to rebase the PR due to conflicts between inside and outside branches.
Could you help us? Thanks a lot.

zejun-chen committed:
Merge branch 'main' into zejun/add_xpu_post_processing_for_torch_profile

zejun-chen commented:
@pytorchbot merge

gujinghui commented:
@pytorchbot rebase

zejun-chen committed:
[Fix] Fix profiler error in torch/autograd/profiler_legacy.py
and torch/testing/_internal/distributed/rpc/rpc_test.py for
EventList attributes in init function

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

aaronenyeshi commented:
Can you please manually rebase on the latest `refs/remotes/origin/viable/strict` branch? I can help import this diff again, but seems there is a rebase conflict that needs to be resolved first.

aaronenyeshi commented:
@pytorchbot rebase -b main

aaronenyeshi committed:
Merge remote-tracking branch 'origin/main' into zejun/add_xpu_post_processing_for_torch_profile

aaronenyeshi committed:
Fix diverging use_device logic, handle privateuse1 as device == privateuseone

aaronenyeshi committed:
Add missing cpu_time_total

aaronenyeshi commented:
I've tried to fix the common errors, and merged to latest main. Let's see how the CI goes.

zejun-chen commented:
> I've tried to fix the common errors, and merged to latest main. Let's see how the CI goes.

Thank you for help. Really appreciate that. 👍 @aaronenyeshi 

While the CI has a few errors:
A. `use_device` maybe None when profiling cpu operation. Maybe we can check it if it is None.
![image](https://github.com/pytorch/pytorch/assets/74231238/87aedbea-be2b-4cf5-93db-e1ba08ce01b3)
https://github.com/pytorch/pytorch/pull/123247/commits/5315fd20bb6ae6aeed3d29a2672477a7de27854f#diff-25291b895a9131f22bf0d5c77ad3403629990bbe9c410256f571f26ce126f461R882

B. the `device_time` is property method, may be we can call it by `return self.device_time`
![image](https://github.com/pytorch/pytorch/assets/74231238/d8c116f7-55b3-4fd0-a2f4-bff541cff0f9)
https://github.com/pytorch/pytorch/pull/123247/commits/5315fd20bb6ae6aeed3d29a2672477a7de27854f#diff-25291b895a9131f22bf0d5c77ad3403629990bbe9c410256f571f26ce126f461R418

aaronenyeshi committed:
Fix CI bugs related to use_device=None, device_time property method, and privateuseone

aaronenyeshi committed:
Fix lints for RUFF and incorrect cuda_memory_usage

aaronenyeshi committed:
Fix docstring indentations and missing cuda_memory_usage

zejun-chen commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Broken trunk" -c ignoredsignal

DanilBaibak commented:
Hi @zejun-chen! Sorry, I need to revert your PR. Your changes broke the [TensorPipeCudaRpcTest::test_profiler_remote_cuda](https://github.com/pytorch/pytorch/actions/runs/8744235568/job/23997655910#step:19:2678) test.

zejun-chen commented:
> Hi @zejun-chen! Sorry, I need to revert your PR. Your changes broke the [TensorPipeCudaRpcTest::test_profiler_remote_cuda](https://github.com/pytorch/pytorch/actions/runs/8744235568/job/23997655910#step:19:2678) test.

Thank you for help. We will quickly take a look and check the failed cases.
BTW, when i start merging, the torch CI told me that the failed case is not related.
The failed case is:  _**python test/distributed/rpc/cuda/test_tensorpipe_agent.py -k test_profiler_remote_cuda**_

aaronenyeshi committed:
Merge remote-tracking branch 'refs/remotes/origin/viable/strict' into zejun/add_xpu_post_processing_for_torch_profile

aaronenyeshi commented:
Looks like we're on the final step, please take a look at failing test:
```
2024-04-19T10:51:41.8146877Z FAILED [4.4162s] distributed/rpc/cuda/test_tensorpipe_agent.py::TensorPipeCudaRpcTest::test_profiler_remote_cuda - RuntimeError: Process 1 exited with error code 10 and exception:
2024-04-19T10:51:41.8147068Z Traceback (most recent call last):
2024-04-19T10:51:41.8148024Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 653, in run_test
2024-04-19T10:51:41.8148351Z     getattr(self, test_name)()
2024-04-19T10:51:41.8149282Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 539, in wrapper
2024-04-19T10:51:41.8149437Z     fn()
2024-04-19T10:51:41.8150314Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2755, in wrapper
2024-04-19T10:51:41.8150493Z     method(*args, **kwargs)
2024-04-19T10:51:41.8151519Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 178, in wrapper
2024-04-19T10:51:41.8151705Z     return func(*args, **kwargs)
2024-04-19T10:51:41.8152625Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/dist_utils.py", line 81, in new_test_method
2024-04-19T10:51:41.8152897Z     return_value = old_test_method(self, *arg, **kwargs)
2024-04-19T10:51:41.8154009Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 4617, in test_profiler_remote_cuda
2024-04-19T10:51:41.8154267Z     self.assertGreater(event.device_time_total, 0)
2024-04-19T10:51:41.8154798Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 1244, in assertGreater
2024-04-19T10:51:41.8155075Z     self.fail(self._formatMessage(msg, standardMsg))
2024-04-19T10:51:41.8155533Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 675, in fail
2024-04-19T10:51:41.8155715Z     raise self.failureException(msg)
2024-04-19T10:51:41.8155917Z AssertionError: 0 not greater than 0
2024-04-19T10:51:41.8155938Z 
2024-04-19T10:51:41.8156276Z To execute this test, run the following from the base repo dir:
2024-04-19T10:51:41.8156879Z      python test/distributed/rpc/cuda/test_tensorpipe_agent.py -k test_profiler_remote_cuda
```
I've rebased on stable branch, let's see if it continues to reproduce. Dr. CI said it was unrelated, let's see.

zejun-chen commented:
> It looks like we're on the final step, please take a look at this failing test:
> 
> ```
> 2024-04-19T10:51:41.8146877Z FAILED [4.4162s] distributed/rpc/cuda/test_tensorpipe_agent.py::TensorPipeCudaRpcTest::test_profiler_remote_cuda - RuntimeError: Process 1 exited with error code 10 and exception:
> 2024-04-19T10:51:41.8147068Z Traceback (most recent call last):
> 2024-04-19T10:51:41.8148024Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 653, in run_test
> 2024-04-19T10:51:41.8148351Z     getattr(self, test_name)()
> 2024-04-19T10:51:41.8149282Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 539, in wrapper
> 2024-04-19T10:51:41.8149437Z     fn()
> 2024-04-19T10:51:41.8150314Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2755, in wrapper
> 2024-04-19T10:51:41.8150493Z     method(*args, **kwargs)
> 2024-04-19T10:51:41.8151519Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 178, in wrapper
> 2024-04-19T10:51:41.8151705Z     return func(*args, **kwargs)
> 2024-04-19T10:51:41.8152625Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/dist_utils.py", line 81, in new_test_method
> 2024-04-19T10:51:41.8152897Z     return_value = old_test_method(self, *arg, **kwargs)
> 2024-04-19T10:51:41.8154009Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 4617, in test_profiler_remote_cuda
> 2024-04-19T10:51:41.8154267Z     self.assertGreater(event.device_time_total, 0)
> 2024-04-19T10:51:41.8154798Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 1244, in assertGreater
> 2024-04-19T10:51:41.8155075Z     self.fail(self._formatMessage(msg, standardMsg))
> 2024-04-19T10:51:41.8155533Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 675, in fail
> 2024-04-19T10:51:41.8155715Z     raise self.failureException(msg)
> 2024-04-19T10:51:41.8155917Z AssertionError: 0 not greater than 0
> 2024-04-19T10:51:41.8155938Z 
> 2024-04-19T10:51:41.8156276Z To execute this test, run the following from the base repo dir:
> 2024-04-19T10:51:41.8156879Z      python test/distributed/rpc/cuda/test_tensorpipe_agent.py -k test_profiler_remote_cuda
> ```
> 
> I've merged the stable branch, so let's see if the failure reproduces. Dr. CI said it was unrelated, but let's see.

Hi, @aaronenyeshi @DanilBaibak 

Thank you for help. It do be a bug from my code change here:
https://github.com/pytorch/pytorch/pull/123247/files#diff-25291b895a9131f22bf0d5c77ad3403629990bbe9c410256f571f26ce126f461R561
I merge the 2 functions `cuda_time_total` and `privateuse1_time_total` into one function `device_time_total`. The 'use_device' should be set instead of None for creating `FunctionEvent` when running `torch.autograd.profiler_legacy.profile()`, otherwise the conditional code in above mentioned link is always True, so that is why CI gets the error: `self.assertGreater(event.device_time_total, 0)`.

Thank you.

aaronenyeshi committed:
Fix legacy profiler event, set use_device=cuda for has_cuda() events

zejun-chen commented:
@pytorchbot merge

malfet commented on torch/autograd/profiler.py:
```diff
@@ -233,17 +240,19 @@ def __init__(
-        if self.use_device and self.use_device != _get_privateuse1_backend_name():
-            warn(f"{self.use_device} doesn't support profile.")
+        VALID_DEVICE_OPTIONS = ["cuda", "xpu", "privateuseone"]
+        if self.use_device not in VALID_DEVICE_OPTIONS:
```
This results in an annoying `UserWarning: The None is not a valid device option.` warning whenever CPU profiler is used :( PR is coming

zejun-chen replied:
Thank you @malfet

malfet commented:
Hmm, why this PR added `#suppress-api-compatibility-check`?

zejun-chen commented:
> Hmm, why this PR added `#suppress-api-compatibility-check`?

In this PR, we removed the `cuda_time` from the `FormattedTimesMixin`, then the lint checking failed because it checks the deleted function. Here is the message we got previously, so we added `#suppress-api-compatibility-check`.
```
2024-04-11T18:09:20.0642879Z ##[warning]Function FormattedTimesMixin.cuda_time: function deleted
2024-04-11T18:09:20.0645300Z ##[warning]Function FormattedTimesMixin.privateuse1_time: function deleted
2024-04-11T18:09:20.0646984Z ##[warning]Function FunctionEvent.self_cuda_memory_usage: function deleted
2024-04-11T18:09:20.0648547Z ##[warning]Function FunctionEvent.self_privateuse1_memory_usage: function deleted
2024-04-11T18:09:20.0649998Z ##[warning]Function FunctionEvent.cuda_time_total: function deleted
2024-04-11T18:09:20.0651407Z ##[warning]Function FunctionEvent.self_cuda_time_total: function deleted
2024-04-11T18:09:20.0652911Z ##[warning]Function FunctionEvent.self_privateuse1_time_total: function deleted
2024-04-11T18:09:20.0654335Z ##[warning]Function FunctionEvent.privateuse1_time_total: function deleted
```
